confmap/expand_test.go

```diff
@@ -0,0 +1,476 @@
+			input:  "${env:HOST${env:PORT?os=${env:OS",
+			output: "${env:HOST${env:PORT?os=${env:OS",
+		},
+		{
+			name:   "only_missing_opening",
+			input:  "env:HOST}env:PORT}?os=env:OS}",
+			output: "env:HOST}env:PORT}?os=env:OS}",
+		},
+		{
+			name:   "close_before_open",
+			input:  "env:HOST}${env:PORT",
+			output: "env:HOST}${env:PORT",
+		},
+		{
+			name:   "old_env_var_nested",
+			input:  "${test:localhost:${PORT}}",
+			output: "${test:localhost:${PORT}}",
+		},
+		// Partial expand.
+		{
+			name:   "missing_opening_bracket_first",
+			input:  "env:HOST}${env:PORT}",
+			output: "env:HOST}3044",
+		},
+		{
+			name:   "missing_opening_bracket_last",
+			input:  "${env:HOST}env:PORT}",
+			output: "localhostenv:PORT}",
+		},
+		{
+			name:   "missing_closing_bracket_first",
+			input:  "${env:HOST${env:PORT}",
+			output: "${env:HOST3044",
+		},
+		{
+			name:   "missing_closing_bracket_last",
+			input:  "${env:HOST}${env:PORT",
+			output: "localhost${env:PORT",
+		},
+		{
+			name:   "multiple_missing_open",
+			input:  "env:HOST}env:PORT}?os=${env:OS}",
+			output: "env:HOST}env:PORT}?os=ubuntu",
+		},
+		{
+			name:   "multiple_missing_closing",
+			input:  "${env:HOST${env:PORT?os=${env:OS}",
+			output: "${env:HOST${env:PORT?os=ubuntu",
+		},
+		{
+			name:   "more_closing_brackets",
+			input:  "${env:HOST}}}}}}${env:PORT?os=${env:OS}",
+			output: "localhost}}}}}${env:PORT?os=ubuntu",
+		},
+		{
+			name:   "more_opening_brackets",
+			input:  "${env:HOST}${${${${${env:PORT}?os=${env:OS}",
+			output: "localhost${${${${3044?os=ubuntu",
+		},
+		{
+			name:   "alternating_missing_opening",
+			input:  "env:HOST}${env:PORT}?os=env:OS}&pr=${env:PR}",
+			output: "env:HOST}3044?os=env:OS}&pr=amd",
+		},
+		{
+			name:   "alternating_missing_closing",
+			input:  "${env:HOST${env:PORT}?os=${env:OS&pr=${env:PR}",
+			output: "${env:HOST3044?os=${env:OS&pr=amd",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			provider := newFakeProvider("input", func(context.Context, string, WatcherFunc) (*Retrieved, error) {
+				return NewRetrieved(map[string]any{tt.name: tt.input})
+			})
+
+			envProvider := newFakeProvider("env", func(_ context.Context, uri string, _ WatcherFunc) (*Retrieved, error) {
+				switch uri {
+				case "env:COMPLEX_VALUE":
+					return NewRetrieved([]any{"localhost:3042"})
+				case "env:HOST":
+					return NewRetrieved("localhost")
+				case "env:OS":
+					return NewRetrieved("ubuntu")
+				case "env:PR":
+					return NewRetrieved("amd")
+				case "env:PORT":
+					return NewRetrieved(3044)
+				case "env:INT":
+					return NewRetrieved(1)
+				case "env:INT32":
+					return NewRetrieved(32)
+				case "env:INT64":
+					return NewRetrieved(64)
+				case "env:FLOAT32":
+					return NewRetrieved(float32(3.25))
+				case "env:FLOAT64":
+					return NewRetrieved(float64(6.4))
+				case "env:BOOL":
+					return NewRetrieved(true)
+				}
+				return nil, errors.New("impossible")
+			})
```

Comment:
Consider to move this as a separate func to simplify the reading of the test case.